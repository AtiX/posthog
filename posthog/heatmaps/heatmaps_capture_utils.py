from typing import Any, Dict, List
from uuid import uuid4

Event = Dict[str, Any]

event_properties_to_keep = [
    "$viewport_height",
    "$viewport_width",
    "$session_id",
    "distinct_id",
    "$current_url",
    "token",
]


def extract_heatmap_events(events: List[Event]) -> List[Event]:
    """
    To save on data transfer and simplify the event processing, we piggyback $heatmap events on top of existing events

    """
    heatmap_events: List[Event] = []

    for event in events:
        heatmap_data = event["properties"].pop("$heatmap_data", None)

        if heatmap_data:
            heatmap_event: Dict = {
                "event": "$heatmap",
                "uuid": str(uuid4()),
                "properties": {
                    "$heatmap_data": heatmap_data,
                },
                "timestamp": event.get("timestamp"),
            }

            # Only copy the event data we need:
            for prop in event_properties_to_keep:
                heatmap_event["properties"][prop] = event["properties"].get(prop, None)

            heatmap_events.append(heatmap_event)

    return events + heatmap_events
