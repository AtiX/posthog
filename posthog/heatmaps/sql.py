from django.conf import settings

from posthog.clickhouse.kafka_engine import kafka_engine
from posthog.clickhouse.table_engines import (
    Distributed,
    ReplicationScheme,
    MergeTreeEngine,
)
from posthog.kafka_client.topics import KAFKA_CLICKHOUSE_HEATMAP_EVENTS

HEATMAPS_DATA_TABLE = lambda: "sharded_heatmaps"


"""
We intend to send specific $heatmap events to build a heatmap instead of building from autocapture like the click map
We'll be storing individual clicks per url/team/session
And we'll be querying for those clicks at day level of granularity
And we'll be querying by URL exact or wildcard match
And we'll _sometimes_ be querying by width

We _could_ aggregate this data by day, but we're hoping this will be small/fast enough not to bother
And can always add a materialized view for day (and week?) granularity driven by this data if needed

We only add session_id so that we could offer example sessions for particular clicked areas in the toolbar
"""

KAFKA_HEATMAPS_TABLE_BASE_SQL = """
CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
(
    $session_id VARCHAR,
    team_id Int64,
    timestamp DateTime64(6, 'UTC'),
    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    x Int16,
    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    y Int16,
    -- stored so that in future we can support other resolutions
    scale_factor Int16,
    $viewport_width Int16,
    $viewport_height Int16,
    -- some elements move when the page scrolls, others do not
    $pointer_target_fixed Bool,
    $current_url VARCHAR
) ENGINE = {engine}
"""

HEATMAPS_TABLE_BASE_SQL = """
CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
(
    $session_id VARCHAR,
    team_id Int64,
    timestamp DateTime64(6, 'UTC'),
    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    x Int16,
    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    y Int16,
    -- stored so that in future we can support other resolutions
    scale_factor Int16,
    $viewport_width Int16,
    $viewport_height Int16,
    -- some elements move when the page scrolls, others do not
    $pointer_target_fixed Bool,
    $current_url VARCHAR,
    _timestamp DateTime
) ENGINE = {engine}
"""

HEATMAPS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine("heatmaps", replication_scheme=ReplicationScheme.SHARDED)

HEATMAPS_TABLE_SQL = lambda: (
    HEATMAPS_TABLE_BASE_SQL
    + """
    PARTITION BY toYYYYMM(timestamp)
    -- almost always this is being queried by
    --   * team_id,
    --   * date range,
    --   * URL (maybe matching wild cards),
    --   * width
    -- we'll almost never query this by session id
    -- so from least to most cardinality that's
    ORDER BY (team_id,  toDate(timestamp), $current_url, $viewport_width)
-- I am purposefully not setting index granularity
-- the default is 8192, and we will be loading a lot of data
-- per query, we tend to copy this 512 around the place but
-- i don't think it applies here
"""
).format(
    table_name=HEATMAPS_DATA_TABLE(),
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=HEATMAPS_DATA_TABLE_ENGINE(),
)

KAFKA_HEATMAPS_TABLE_SQL = lambda: KAFKA_HEATMAPS_TABLE_BASE_SQL.format(
    table_name="kafka_heatmaps",
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=kafka_engine(topic=KAFKA_CLICKHOUSE_HEATMAP_EVENTS),
)

HEATMAPS_TABLE_MV_SQL = (
    lambda: """
CREATE MATERIALIZED VIEW IF NOT EXISTS heatmaps_mv ON CLUSTER '{cluster}'
TO {database}.{target_table}
AS SELECT
    $session_id,
    team_id,
    timestamp,
    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    x,
    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
    y,
    -- stored so that in future we can support other resolutions
    scale_factor,
    $viewport_width,
    $viewport_height,
    -- some elements move when the page scrolls, others do not
    $pointer_target_fixed,
    $current_url
FROM {database}.kafka_heatmaps
""".format(
        target_table="writable_heatmaps",
        cluster=settings.CLICKHOUSE_CLUSTER,
        database=settings.CLICKHOUSE_DATABASE,
    )
)

# Distributed engine tables are only created if CLICKHOUSE_REPLICATED

# This table is responsible for writing to sharded_heatmaps based on a sharding key.
WRITABLE_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
    table_name="writable_heatmaps",
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=Distributed(
        data_table=HEATMAPS_DATA_TABLE(),
        # I don't think there's a great natural sharding key here
        # we'll be querying for team data by url and date,
        # so I _think_ this offers a reasonable spread of write load
        # without needing to query too many shards
        sharding_key="cityHash64(concat(toString(team_id), '-', $current_url, '-', toString(toDate(timestamp))))",
    ),
)

# This table is responsible for reading from heatmaps on a cluster setting
DISTRIBUTED_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
    table_name="heatmaps",
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=Distributed(
        data_table=HEATMAPS_DATA_TABLE(),
        # I don't think there's a great natural sharding key here
        # we'll be querying for team data by url and date,
        # so I _think_ this offers a reasonable spread of write load
        # without needing to query too many shards
        sharding_key="cityHash64(concat(toString(team_id), '-', $current_url, '-', toString(toDate(timestamp))))",
    ),
)

DROP_HEATMAPS_TABLE_SQL = lambda: (
    f"DROP TABLE IF EXISTS {HEATMAPS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
)

TRUNCATE_HEATMAPS_TABLE_SQL = lambda: (
    f"TRUNCATE TABLE IF EXISTS {HEATMAPS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
)
