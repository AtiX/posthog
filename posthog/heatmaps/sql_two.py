from django.conf import settings

from posthog.clickhouse.kafka_engine import kafka_engine
from posthog.kafka_client.topics import KAFKA_CLICKHOUSE_HEATMAP_EVENTS

HEATMAPS_DATA_TABLE = lambda: "sharded_heatmaps"

# using JSONAsString we can send multiple JSON items in a single kafka messages,
# but we have to write to a String column
# JSONEachRow checks that the payload it is receiving starts with '{'
KAFKA_HEATMAPS_TABLE_BASE_SQL = """
CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
(
    kafka_payload String
) ENGINE = {engine}
"""

KAFKA_HEATMAPS_TWO_TABLE_SQL = lambda: KAFKA_HEATMAPS_TABLE_BASE_SQL.format(
    table_name="kafka_heatmaps_two",
    cluster=settings.CLICKHOUSE_CLUSTER,
    engine=kafka_engine(topic=KAFKA_CLICKHOUSE_HEATMAP_EVENTS, serialization="JSONAsString"),
)

# now the materialized view that ingests from it "simply" extracts values from that json payload
HEATMAPS_TABLE_TWO_MV_SQL = (
    lambda: """
CREATE MATERIALIZED VIEW IF NOT EXISTS heatmaps_two_mv ON CLUSTER '{cluster}'
TO {database}.{target_table}
AS SELECT
    JSONExtractString(kafka_payload, 'session_id') as session_id,
    JSONExtractInt(kafka_payload, 'team_id') as team_id,
    JSONExtractString(kafka_payload, 'distinct_id') as distinct_id,
    toDateTime64(JSONExtractString(kafka_payload, 'timestamp'), 6, 'UTC') AS timestamp,
    JSONExtractInt(kafka_payload, 'x') AS x,
    JSONExtractInt(kafka_payload, 'y') AS y,
    JSONExtractInt(kafka_payload, 'scale_factor') AS scale_factor,
    JSONExtractInt(kafka_payload, 'viewport_width') AS viewport_width,
    JSONExtractInt(kafka_payload, 'viewport_height') AS viewport_height,
    JSONExtractBool(kafka_payload, 'pointer_target_fixed') AS pointer_target_fixed,
    JSONExtractString(kafka_payload, 'current_url') AS current_url,
    JSONExtractString(kafka_payload, 'type') AS type,
    _timestamp,
    _offset,
    _partition
FROM {database}.kafka_heatmaps_two
""".format(
        target_table="writable_heatmaps",
        cluster=settings.CLICKHOUSE_CLUSTER,
        database=settings.CLICKHOUSE_DATABASE,
    )
)
