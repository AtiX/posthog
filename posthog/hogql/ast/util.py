def parse_string_literal(ctx):
    text = ctx.getText()
    text = text[1:-1]
    # from clickhouse's grammar
    text = text.replace("''", "'")
    # copied from clickhouse_driver/util/escape.py
    text = text.replace("\\b", "\b")
    text = text.replace("\\f", "\f")
    text = text.replace("\\r", "\r")
    text = text.replace("\\n", "\n")
    text = text.replace("\\t", "\t")
    text = text.replace("\\0", "\0")
    text = text.replace("\\a", "\a")
    text = text.replace("\\v", "\v")
    text = text.replace("\\\\", "\\")
    text = text.replace("\\'", "'")

    return text
