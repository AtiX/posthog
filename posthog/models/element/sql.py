CREATE_ELEMENTS_CHAIN_DAILY_COUNTS_TABLE = """
CREATE TABLE elements_chain_counts_daily
(
    day            DateTime64(3, 'UTC'),
    team_id        int,
    elements_chain String,
    "$current_url" String,
    count          int
)
ENGINE = SummingMergeTree
PARTITION BY toYYYYMM(day) ORDER BY (team_id,day,"$current_url", elements_chain);
"""

CREATE_ELEMENTS_CHAIN_DAILY_COUNTS_MV = """
CREATE MATERIALIZED VIEW elements_chain_counts_daily_mv
TO elements_chain_counts_daily
AS SELECT
    toStartOfDay(timestamp) as day,
    team_id,
    elements_chain,
    replaceRegexpAll(JSONExtractRaw(properties, '$current_url'), '^"|"$', '') as "$current_url",
    count(*) as count
FROM events
WHERE event = '$autocapture'
AND elements_chain != ''
GROUP BY team_id, day, elements_chain, "$current_url";
"""

GET_ELEMENTS_FROM_MV = """
SELECT
    elements_chain, sum(count) as count
FROM elements_chain_counts_daily_mv
WHERE
    team_id = %(team_id)s
    AND day >= toStartOfDay(%(date_from)s, 'UTC')
    AND day <= %(date_to)s
    AND "$current_url" = %(current_url)s
GROUP BY elements_chain
ORDER BY count DESC
"""


GET_ELEMENTS = """
SELECT
    elements_chain, count(1) as count
FROM events
WHERE
    team_id = %(team_id)s AND
    event = '$autocapture' AND
    elements_chain != ''
    {date_from}
    {date_to}
    {query}
GROUP BY elements_chain
ORDER BY count DESC
LIMIT {limit}{conditional_offset};
"""

GET_VALUES = """
SELECT
    extract(elements_chain, %(regex)s) as value, count(1) as count
FROM (
      SELECT elements_chain
      FROM events
      WHERE team_id = %(team_id)s
        AND event = '$autocapture'
        AND elements_chain != ''
        AND match(elements_chain, %(filter_regex)s)
        LIMIT 100000
)
GROUP BY value
ORDER BY count desc
LIMIT 100;
"""
