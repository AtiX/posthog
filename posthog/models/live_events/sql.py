from posthog.clickhouse.kafka_engine import kafka_engine
from posthog.clickhouse.table_engines import Distributed, MergeTreeEngine, ReplicationScheme
from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE, KAFKA_EVENTS_PLUGIN_INGESTION_TOPIC

LIVE_EVENTS_TABLE_BASE_SQL = """
CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
(
    uuid UUID,
    event VARCHAR,
    properties VARCHAR,
    timestamp DateTime64(6, 'UTC'),
    team_id Int64,
    distinct_id VARCHAR,
    elements_chain VARCHAR,
    created_at DateTime64(6, 'UTC'),
    _timestamp Nullable(DateTime)
) ENGINE = {engine}
"""


LIVE_EVENTS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine(
    "sharded_live_events", replication_scheme=ReplicationScheme.SHARDED
)

# TODO: Double check TTL and if we want to manually drop partitions e.g. every hour
LIVE_EVENTS_DATA_TABLE_SQL = lambda: (
    LIVE_EVENTS_TABLE_BASE_SQL
    + """PARTITION BY toStartOfTenMinutes(timestamp)
ORDER BY (team_id, timestamp, event, cityHash64(distinct_id), cityHash64(uuid))
TTL toDateTime(timestamp) + INTERVAL 1 DAY
"""
).format(
    table_name="sharded_live_events",
    cluster=CLICKHOUSE_CLUSTER,
    engine=LIVE_EVENTS_DATA_TABLE_ENGINE(),
)

# This table is responsible for writing to sharded_live_events based on a sharding key
# We also read from this table instead of sharded_live_events
DISTRIBUTED_LIVE_EVENTS_TABLE_SQL = lambda: LIVE_EVENTS_TABLE_BASE_SQL.format(
    table_name="live_events",
    cluster=CLICKHOUSE_CLUSTER,
    engine=Distributed(data_table="sharded_live_events", sharding_key="rand()"),  # use team_id?
)


KAFKA_LIVE_EVENTS_TABLE_SQL = lambda: LIVE_EVENTS_TABLE_BASE_SQL.format(
    table_name="kafka_live_events",
    cluster=CLICKHOUSE_CLUSTER,
    # being explicit about the group even though we're using the default
    # to highlight that this is important - we should never use the same
    # consumer group as the plugin server here!
    engine=kafka_engine(topic=KAFKA_EVENTS_PLUGIN_INGESTION_TOPIC, group="group1"),
)

# KLUDGE: We don't have the real timestamp yet, so we use now()
# currently this is handled in the plugin server through the
# handleTimestamp function, but we should port it to Django
#
# Note: we use 0 as created_at because the event hasn't been created yet,
# and that actually allows us to correctly aggregate arguments when
# doing a merge query
LIVE_EVENTS_MV_TABLE_SQL = lambda: """
CREATE MATERIALIZED VIEW live_events_mv ON CLUSTER '{cluster}'
TO {database}.{target_table}
AS SELECT
    uuid,
    event,
    properties,
    toDateTime64(now(), 6, 'UTC') as timestamp,
    team_id,
    distinct_id,
    '' as elements_chain,
    toDateTime64(0, 6, 'UTC') as created_at,
    _timestamp
FROM {database}.kafka_live_events
WHERE event != '$snapshot'
""".format(
    target_table="live_events",
    cluster=CLICKHOUSE_CLUSTER,
    database=CLICKHOUSE_DATABASE,
)


SELECT_LIVE_EVENTS_BY_TEAM_AND_CONDITIONS_FILTERS_SQL = """
SELECT
    uuid,
    event,
    team_id,
    distinct_id,
    properties_aggregated as properties,
    timestamp_aggregated as timestamp,
    elements_chain_aggregated as elements_chain,
    source_table
FROM (
    SELECT
        uuid,
        event,
        team_id,
        distinct_id,
        argMax(properties, created_at) as properties_aggregated,
        argMax(elements_chain, created_at) as elements_chain_aggregated,
        argMax(timestamp, created_at) as timestamp_aggregated,
        argMax(_table, created_at) as source_table
    FROM
        merge('{database}', '^(events|live_events)$')
    WHERE
        team_id = %(team_id)s
        {conditions}
        {filters}
    GROUP BY uuid, event, team_id, distinct_id
    ORDER BY timestamp_aggregated {order} {limit}
)
"""


SELECT_LIVE_EVENTS_BY_TEAM_AND_CONDITIONS_SQL = """
SELECT
    uuid,
    event,
    team_id,
    distinct_id,
    properties_aggregated as properties,
    timestamp_aggregated as timestamp,
    elements_chain_aggregated as elements_chain,
    source_table
FROM (
    SELECT
        uuid,
        event,
        team_id,
        distinct_id,
        argMax(properties, created_at) as properties_aggregated,
        argMax(elements_chain, created_at) as elements_chain_aggregated,
        argMax(timestamp, created_at) as timestamp_aggregated,
        argMax(_table, created_at) as source_table
    FROM
        merge('{database}', '^(events|live_events)$')
    WHERE
        team_id = %(team_id)s
        {conditions}
    GROUP BY uuid, event, team_id, distinct_id
    ORDER BY timestamp_aggregated {order} {limit}
)
"""

# test utilities
INSERT_INTO_LIVE_EVENTS_SQL = """
INSERT INTO live_events (
    uuid,
    event,
    properties,
    timestamp,
    team_id,
    distinct_id,
    elements_chain,
    created_at,
    _timestamp
)
VALUES (
    %(uuid)s,
    %(event)s,
    %(properties)s,
    %(timestamp)s,
    %(team_id)s,
    %(distinct_id)s,
    '',
    toDateTime64(0, 6, 'UTC'),
    now()
)
"""
