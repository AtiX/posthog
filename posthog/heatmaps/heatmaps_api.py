from datetime import datetime, timedelta
from typing import Any

from rest_framework import viewsets, request, response, serializers, status

from posthog.api.routing import TeamAndOrgViewSetMixin
from posthog.auth import TemporaryTokenAuthentication
from posthog.hogql.ast import Constant
from posthog.hogql.base import Expr
from posthog.hogql.query import execute_hogql_query
from posthog.rate_limit import ClickHouseSustainedRateThrottle, ClickHouseBurstRateThrottle


class HeatmapResponseItemSerializer(serializers.Serializer):
    count = serializers.IntegerField(required=True)
    pointer_y = serializers.IntegerField(required=True)
    pointer_relative_x = serializers.FloatField(required=True)
    pointer_target_fixed = serializers.BooleanField(required=True)


def default_start_date():
    return (datetime.now() - timedelta(days=7)).date()


class HeatmapsRequestSerializer(serializers.Serializer):
    viewport_min_width = serializers.IntegerField(required=False)
    viewport_max_width = serializers.IntegerField(required=False)
    type = serializers.CharField(required=False)
    date_from = serializers.DateField(required=False, default=default_start_date)
    date_to = serializers.DateField(required=False)
    url_exact = serializers.CharField(required=False)
    url_pattern = serializers.CharField(required=False)


class HeatmapsResponseSerializer(serializers.Serializer):
    results = HeatmapResponseItemSerializer(many=True)


class HeatmapViewSet(TeamAndOrgViewSetMixin, viewsets.GenericViewSet):
    scope_object = "INTERNAL"
    filter_rewrite_rules = {"team_id": "group__team_id"}

    throttle_classes = [ClickHouseBurstRateThrottle, ClickHouseSustainedRateThrottle]
    serializer_class = HeatmapsResponseSerializer

    authentication_classes = [TemporaryTokenAuthentication]

    def get_queryset(self):
        return None

    def list(self, request: request.Request, *args: Any, **kwargs: Any) -> response.Response:
        request_serializer = HeatmapsRequestSerializer(data=request.query_params)
        request_serializer.is_valid(raise_exception=True)

        placeholders: dict[str, Expr] = {
            "team_id": Constant(value=self.team.pk),
            "date_from": Constant(value=request_serializer.validated_data["date_from"]),
        }
        if request_serializer.validated_data.get("viewport_min_width", None):
            placeholders["vp_min_w"] = Constant(value=request_serializer.validated_data["viewport_min_width"])
        if request_serializer.validated_data.get("viewport_max_width", None):
            placeholders["vp_max_w"] = Constant(value=request_serializer.validated_data["viewport_max_width"])
        if request_serializer.validated_data.get("date_to", None):
            placeholders["date_to"] = Constant(value=request_serializer.validated_data["date_to"])
        if request_serializer.validated_data.get("url_exact", None):
            placeholders["url_exact"] = Constant(value=request_serializer.validated_data["url_exact"])
        if request_serializer.validated_data.get("url_pattern", None):
            placeholders["url_pattern"] = Constant(value=request_serializer.validated_data["url_pattern"])

        q = """
            select *, count() as cnt
            from (
                     select pointer_target_fixed, round((x / viewport_width), 2) as relative_client_x,
                            y * scale_factor                  as client_y
                     from heatmaps
                     where 1=1
                     {date_from_predicate}
                    {date_to_predicate}
                    {viewport_min_width_predicate}
                    {viewport_max_width_predicate}
                    {url_exact_predicate}
                    {url_pattern_predicate}
                     {team_id_predicate}
                     )
            group by `pointer_target_fixed`, relative_client_x, client_y
            """.format(
            # required
            date_from_predicate="and timestamp >= {date_from}",
            team_id_predicate="and team_id = {team_id}",
            # optional
            date_to_predicate="and timestamp <= {date_to} + interval 1 day"
            if request_serializer.validated_data.get("date_to", None)
            else "",
            viewport_min_width_predicate="and viewport_width >= ceil({vp_min_w} / 16)"
            if request_serializer.validated_data.get("viewport_min_width", None)
            else "",
            viewport_max_width_predicate="and viewport_width <= ceil({vp_max_w} / 16)"
            if request_serializer.validated_data.get("viewport_max_width", None)
            else "",
            url_exact_predicate="and current_url = {url_exact}"
            if request_serializer.validated_data.get("url_exact", None)
            else "",
            url_pattern_predicate="and current_url like {url_pattern}"
            if request_serializer.validated_data.get("url_pattern", None)
            else "",
        )

        doohickies = execute_hogql_query(
            query=q,
            placeholders=placeholders,
            team=self.team,
        )

        data = [
            {
                "pointer_target_fixed": item[0],
                "pointer_relative_x": item[1],
                "pointer_y": item[2],
                "count": item[3],
            }
            for item in doohickies.results or []
        ]

        response_serializer = HeatmapsResponseSerializer(data={"results": data})
        response_serializer.is_valid(raise_exception=True)
        return response.Response(response_serializer.data, status=status.HTTP_200_OK)


class LegacyHeatmapViewSet(HeatmapViewSet):
    derive_current_team_from_user_only = True
