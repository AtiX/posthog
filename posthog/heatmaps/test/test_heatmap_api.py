import math
from typing import Dict

from posthog.heatmaps.sql import INSERT_SINGLE_HEATMAP_EVENT
from posthog.kafka_client.client import ClickhouseProducer
from posthog.kafka_client.topics import KAFKA_CLICKHOUSE_SESSION_REPLAY_EVENTS
from posthog.models.event.util import format_clickhouse_timestamp
from posthog.test.base import APIBaseTest, ClickhouseTestMixin, QueryMatchingTest, snapshot_clickhouse_queries


class TestSessionRecordings(APIBaseTest, ClickhouseTestMixin, QueryMatchingTest):
    CLASS_DATA_LEVEL_SETUP = False

    def _assert_heatmap_single_result_count(self, params: Dict[str, str] | None, expected_grouped_count: int) -> None:
        if params is None:
            params = {}

        query_params = "&".join([f"{key}={value}" for key, value in params.items()])

        response = self.client.get(f"/api/heatmap/?{query_params}")
        assert response.status_code == 200
        assert len(response.json()["results"]) == 1
        assert response.json()["results"][0]["count"] == expected_grouped_count

    @snapshot_clickhouse_queries
    def test_can_get_empty_response(self) -> None:
        response = self.client.get("/api/heatmap/")
        assert response.status_code == 200
        self.assertEqual(response.json(), {"results": []})

    @snapshot_clickhouse_queries
    def test_can_get_all_data_response(self) -> None:
        self._create_heatmap_event("session_1", "click")
        self._create_heatmap_event("session_2", "click")

        self._assert_heatmap_single_result_count({"date_from": "2023-03-08"}, 2)

    @snapshot_clickhouse_queries
    def test_can_get_filter_by_date_from(self) -> None:
        self._create_heatmap_event("session_1", "click", "2023-03-07T07:00:00")
        self._create_heatmap_event("session_2", "click", "2023-03-08T08:00:00")

        self._assert_heatmap_single_result_count({"date_from": "2023-03-08"}, 1)

    @snapshot_clickhouse_queries
    def test_can_get_filter_by_click(self) -> None:
        self._create_heatmap_event("session_1", "click", "2023-03-08T07:00:00")
        self._create_heatmap_event("session_2", "rageclick", "2023-03-08T08:00:00")
        self._create_heatmap_event("session_2", "rageclick", "2023-03-08T08:01:00")

        self._assert_heatmap_single_result_count({"date_from": "2023-03-08", "type": "click"}, 1)

        self._assert_heatmap_single_result_count({"date_from": "2023-03-08", "type": "rageclick"}, 2)

    # @snapshot_clickhouse_queries
    # def test_can_get_filter_by_viewport(self) -> None:
    #     self._create_heatmap_event("session_1", "click", "2023-03-08T08:00:00", 150)
    #     self._create_heatmap_event("session_2", "rageclick", "2023-03-08T08:00:00", 151)
    #     self._create_heatmap_event("session_2", "rageclick", "2023-03-08T08:01:00", 152)
    #
    #     self._assert_heatmap_single_result_count({"date_from": "2023-03-08", "viewport_width_min": "150"}, 3)
    #     self._assert_heatmap_single_result_count({"date_from": "2023-03-08", "viewport_width_min": "151"}, 1)
    #     self._assert_heatmap_single_result_count({"date_from": "2023-03-08", "viewport_width_min": "152"}, 1)
    #     self._assert_heatmap_single_result_count({"date_from": "2023-03-08", "viewport_width_min": "153"}, 1)

    def _create_heatmap_event(
        self, session_id: str, type: str, date_from: str = "2023-03-08T09:00:00", viewport_width: int = 100
    ) -> None:
        p = ClickhouseProducer()
        # because this is in a test it will write directly using SQL not really with Kafka
        p.produce(
            topic=KAFKA_CLICKHOUSE_SESSION_REPLAY_EVENTS,
            sql=INSERT_SINGLE_HEATMAP_EVENT,
            data={
                "session_id": session_id,
                "team_id": self.team.pk,
                "timestamp": format_clickhouse_timestamp(date_from),
                "x": 10,
                "y": 20,
                "scale_factor": 16,
                # this adjustment is done at ingestion
                "viewport_width": math.ceil(viewport_width / 16),
                "viewport_height": math.ceil(100 / 16),
                "type": type,
                "pointer_target_fixed": True,
                "current_url": "http://example.com",
            },
        )
