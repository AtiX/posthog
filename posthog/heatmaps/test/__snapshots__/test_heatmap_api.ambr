# serializer version: 1
# name: TestSessionRecordings.test_can_get_all_data_response
  '''
  /* user_id:0 request:_snapshot_ */
  SELECT pointer_target_fixed AS pointer_target_fixed,
         relative_client_x AS relative_client_x,
         client_y AS client_y,
         count() AS cnt
  FROM
    (SELECT heatmaps.pointer_target_fixed AS pointer_target_fixed,
            round(divide(heatmaps.x, heatmaps.viewport_width), 2) AS relative_client_x,
            multiply(heatmaps.y, heatmaps.scale_factor) AS client_y
     FROM heatmaps
     WHERE and(equals(heatmaps.team_id, 2), 1, greaterOrEquals(toTimeZone(heatmaps.timestamp, 'UTC'), toDate('2023-03-08')), equals(heatmaps.type, 'click'), equals(heatmaps.team_id, 2)))
  GROUP BY pointer_target_fixed,
           relative_client_x,
           client_y
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1
  '''
# ---
# name: TestSessionRecordings.test_can_get_empty_response
  '''
  /* user_id:0 request:_snapshot_ */
  SELECT pointer_target_fixed AS pointer_target_fixed,
         relative_client_x AS relative_client_x,
         client_y AS client_y,
         count() AS cnt
  FROM
    (SELECT heatmaps.pointer_target_fixed AS pointer_target_fixed,
            round(divide(heatmaps.x, heatmaps.viewport_width), 2) AS relative_client_x,
            multiply(heatmaps.y, heatmaps.scale_factor) AS client_y
     FROM heatmaps
     WHERE and(equals(heatmaps.team_id, 2), 1, greaterOrEquals(toTimeZone(heatmaps.timestamp, 'UTC'), toDate('2024-04-05')), equals(heatmaps.type, 'click'), equals(heatmaps.team_id, 2)))
  GROUP BY pointer_target_fixed,
           relative_client_x,
           client_y
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1
  '''
# ---
# name: TestSessionRecordings.test_can_get_filter_by_click
  '''
  /* user_id:0 request:_snapshot_ */
  SELECT pointer_target_fixed AS pointer_target_fixed,
         relative_client_x AS relative_client_x,
         client_y AS client_y,
         count() AS cnt
  FROM
    (SELECT heatmaps.pointer_target_fixed AS pointer_target_fixed,
            round(divide(heatmaps.x, heatmaps.viewport_width), 2) AS relative_client_x,
            multiply(heatmaps.y, heatmaps.scale_factor) AS client_y
     FROM heatmaps
     WHERE and(equals(heatmaps.team_id, 2), 1, greaterOrEquals(toTimeZone(heatmaps.timestamp, 'UTC'), toDate('2023-03-08')), equals(heatmaps.type, 'click'), equals(heatmaps.team_id, 2)))
  GROUP BY pointer_target_fixed,
           relative_client_x,
           client_y
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1
  '''
# ---
# name: TestSessionRecordings.test_can_get_filter_by_click.1
  '''
  /* user_id:0 request:_snapshot_ */
  SELECT pointer_target_fixed AS pointer_target_fixed,
         relative_client_x AS relative_client_x,
         client_y AS client_y,
         count() AS cnt
  FROM
    (SELECT heatmaps.pointer_target_fixed AS pointer_target_fixed,
            round(divide(heatmaps.x, heatmaps.viewport_width), 2) AS relative_client_x,
            multiply(heatmaps.y, heatmaps.scale_factor) AS client_y
     FROM heatmaps
     WHERE and(equals(heatmaps.team_id, 2), 1, greaterOrEquals(toTimeZone(heatmaps.timestamp, 'UTC'), toDate('2023-03-08')), equals(heatmaps.type, 'rageclick'), equals(heatmaps.team_id, 2)))
  GROUP BY pointer_target_fixed,
           relative_client_x,
           client_y
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1
  '''
# ---
# name: TestSessionRecordings.test_can_get_filter_by_date_from
  '''
  /* user_id:0 request:_snapshot_ */
  SELECT pointer_target_fixed AS pointer_target_fixed,
         relative_client_x AS relative_client_x,
         client_y AS client_y,
         count() AS cnt
  FROM
    (SELECT heatmaps.pointer_target_fixed AS pointer_target_fixed,
            round(divide(heatmaps.x, heatmaps.viewport_width), 2) AS relative_client_x,
            multiply(heatmaps.y, heatmaps.scale_factor) AS client_y
     FROM heatmaps
     WHERE and(equals(heatmaps.team_id, 2), 1, greaterOrEquals(toTimeZone(heatmaps.timestamp, 'UTC'), toDate('2023-03-08')), equals(heatmaps.type, 'click'), equals(heatmaps.team_id, 2)))
  GROUP BY pointer_target_fixed,
           relative_client_x,
           client_y
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1
  '''
# ---
